ponix4-src/source/xorg-ponix4/app/transset-1.0.0/man/transset.1
.ds q \N'34'
.TH TRANSSET 1 "transset 1.0.0" "X Version 11"
.SH NAME
transset \- Set transparency on a window
.SH SYNOPSIS
.BI "transset [\-d " display "] n
.SH DESCRIPTION
.B transset
is a simple program for X servers supporting the XFIXES, DAMAGE,
and COMPOSITE extensions.  It lets the user set the transparency on a window.
.SH OPTIONS
.TP
.BI n
Where n is a number from 0 to 1, zero being transparent and 1 being opaque
.SH BUGS
Probably.  Please report any you find to https://bugs.freedesktop.org/enter_bug.cgi?product=xorg.
.SH AUTHORS
Matthew Hawn


ponix4-const/history.txt
2011-02-03 19:45 - 2.00.575-beta - added graphical libs to usb-ip package
2011-02-04 14:45 - 2.00.576-beta -
2011-02-07 18:10 - 2.00.577-beta - build MTRR (vesa not working)
2011-02-07 18:19 - 2.00.578-beta -
2011-02-10 20:02 - 2.00.579-beta - libsane 1.0.21 with avision backend from git
2011-02-25 18:09 - 2.00.580-beta - edit libsane package
2011-02-25 18:16 - 2.00.581-beta -
2011-02-27 19:50 - 2.00.582-beta - usb-server package first build. included like package
2011-02-28 21:00 - 2.00.583-beta - final release
2011-02-28 21:07 - 2.00.584 - final release
2011-02-28 21:21 - 2.00.585 - see previos message
2011-02-28 21:45 - 2.1.001 - new version format
2011-02-28 22:51 - 2.1.1 - new version
2011-02-28 23:38 - 2.1.2 - sane package
2011-03-01 20:50 - 2.1.3 - new package sane 1.0.22
2011-03-01 23:05 - 2.1.4 -
2011-03-31 22:06 - 2.1.5 - lzma for package and usb2com converter
2011-04-18 16:44 - 2.1.6 - rxvt.pkg
2011-04-18 18:33 - 2.1.7 - rxvt.pkg ver2
2011-04-18 18:41 - 2.1.8 - rxvt ver3
2011-05-06 15:25 - 2.1.9 - repair freerdp server connect
2011-06-29 21:06 - 2.1.10 - old ver
2011-06-29 23:52 - 2.1.11 - intel video driver 2.15.0, freerdp from git, package rdesktop 1.7
2011-06-30 00:06 - 2.1.12 -
2011-06-30 00:34 - 2.1.13 -
2011-06-30 13:27 - 2.1.14 -
2011-06-30 14:20 - 2.1.15 -
2011-06-30 15:14 - 2.1.16 -
2011-06-30 15:40 - 2.1.17 -
2011-06-30 16:04 - 2.1.18 -
2011-07-01 00:11 - 2.1.19 -
2011-07-01 01:15 - 2.1.20 -
2011-07-05 20:51 - 2.1.21-standalone - rename to standalone version
2011-07-06 14:16 - 2.1.22 - add intel driver 2.15.0 and last version libdrm
2011-07-06 18:37 - 2.1.23 - rdesktop+keymaps
2011-07-07 16:05 - 2.1.24 - add keymaps to the freerdp17 package
2011-07-07 19:16 - 2.1.25 - error in the rdesktop package
2011-07-07 19:50 - 2.1.26 - too
2011-07-07 20:10 - 2.1.27 - vmview auto start
2011-07-07 20:16 - 2.1.28 - too
2011-07-07 21:08 - 2.1.29 - edit vmopenclient
2011-07-08 00:11 - 2.1.30 - edit vmopenclient
2011-07-12 21:18 - 2.1.31 - etoken
2011-07-13 18:28 - 2.1.32 - repair sane package
2011-07-13 19:02 - 2.1.33 - add smart-card support to rdesktop17 package
2011-07-21 13:38 - 2.1.34 - last version of realtek ethernet driver r8101
2011-07-21 14:52 - 2.1.35 - add firmware fo rtl8105e
2011-07-21 19:18 - 2.1.36 - new version of realtek r8169 driver (15)
2011-07-21 20:55 - 2.1.37 - without r8101.ko
2011-07-21 21:27 - 2.1.38 - r8169 from kernek, r8101 delete
2011-07-22 16:45 - 2.1.39 - add citrix client package
2011-07-22 16:51 - 2.1.40 -
2011-07-22 19:03 - 2.1.41 -
2011-07-22 19:59 - 2.1.42 -
2011-07-22 20:07 - 2.1.43 -
2011-07-22 20:40 - 2.1.44 -
2011-07-23 14:11 - 2.1.45-standalone - trying wifi build with odinary /etc/init.d/network (no ethtool)
2011-07-26 13:16 - 2.1.46 -
2011-07-26 15:20 - 2.1.47 -
2011-07-26 16:22 - 2.1.48 -
2011-07-27 14:30 - 2.1.49 -
2011-07-27 15:43 - 2.1.50 -
2011-07-27 18:05 - 2.1.51 -
2011-08-01 13:59 - 2.1.52 -
2011-08-01 15:36 - 2.1.53 -
2011-08-01 16:25 - 2.1.54 -
2011-08-01 18:18 - 2.1.55 -
2011-08-01 20:09 - 2.1.56 -
2011-08-01 20:10 - 2.1.57 - renew xorg server and libXfont
2011-08-01 20:43 - 2.1.58 -
2011-08-01 21:40 - 2.1.59 -
2011-08-01 22:32 - 2.1.60 -
2011-08-01 23:24 - 2.1.61 -
2011-08-01 23:37 - 2.1.62 -
2011-08-02 13:53 - 2.1.63 -
2011-08-02 15:07 - 2.1.64 -
2011-08-02 19:31 - 2.1.65 -
2011-08-02 20:30 - 2.1.66 -
2011-08-02 21:16 - 2.1.67 -
2011-08-02 22:29 - 2.1.68 -
2011-08-02 22:37 - 2.1.69 -
2011-08-03 16:08 - 2.1.70 -
2011-08-03 19:12 - 2.1.71 -
2011-08-03 19:47 - 2.1.72 - ICA 11 packages working!
2011-08-03 20:34 - 2.1.73 -
2011-08-03 20:43 - 2.1.74 -
2011-08-10 22:32 - 2.1.75 - add no_tls and no_nla to freerdp parameters
2011-08-11 00:30 - 2.1.76-standalone -
2011-08-11 18:35 - 2.1.77 - repair bug with freerdp_no_nla freerdp_no_tls
2011-08-11 21:45 - 2.1.78 - repair bug with Xorg (absent libfreetype)
2011-08-12 15:15 - 2.1.79 - add relatime parameter to udev script
2011-08-16 15:07 - 2.1.80 -
2011-08-16 17:26 - 2.1.81 -
2011-08-16 17:51 - 2.1.82 -
2011-08-16 21:43 - 2.1.83 - add gstreamer for ica11 package
2011-08-16 22:31 - 2.1.84 - add gstreamer and plugins for ica11 package
2011-08-18 13:57 - 2.1.85 - add gstream for speedscreen function for citrix client
2011-08-18 14:47 - 2.1.86 - add oss support to the alsa drivers
2011-08-18 15:11 - 2.1.87 -
2011-08-18 16:08 - 2.1.88 - rebuild libasound
2011-08-18 18:01 - 2.1.89 - error, soundcore must be from the kernel
2011-08-18 18:55 - 2.1.90 -
2011-09-12 18:06 - 2.1.91 - correct error with rdesktop17 package add pcsclite library
2011-09-12 22:31 - 2.1.92 - add icewm
2011-09-12 23:00 - 2.1.93 - add graphical libs to icewm
2011-09-12 23:11 - 2.1.94 - add locales to icewm
2011-09-12 23:37 - 2.1.95 -
2011-09-12 23:45 - 2.1.96 -
2011-09-13 00:07 - 2.1.97 -
2011-09-13 00:22 - 2.1.98 -
2011-09-13 13:32 - 2.1.99 -
2011-09-13 14:17 - 2.1.100 - 1
2011-09-13 19:30 - 2.1.101 -
2011-09-13 20:01 - 2.1.102 -
2011-09-13 20:40 - 2.1.103 -
2011-09-13 23:41 - 2.1.104 -
2011-09-14 15:00 - 2.1.105 -
2011-09-14 15:37 - 2.1.106 -
2011-09-16 15:08 - 2.1.107 -
2011-09-26 21:38 - 2.1.108 - in network_script added SERVER_IP=$serverid. hack to use dhcp server as tftp when no siaddr specified (current udhcpc has a bug and doesn't set tftp option)
2011-09-27 20:49 - 2.1.109 - last version of nvidia xorg driver
2011-09-28 14:48 - 2.1.110 - add --no-nla parameter for freerdp client
2011-10-03 20:00 - 2.1.111 - new version of atheros ethernet driver atl1e (AR81Family-linux-1.0.1.14) instead atl1c and atl1e (now removed)
2011-10-03 20:24 - 2.1.112-standalone - current update
2011-10-03 20:52 - 2.1.113 -
2011-10-03 21:22 - 2.1.114 -
2011-10-03 22:10 - 2.1.115 - udev-debug
2011-10-03 22:56 - 2.1.116 - udev debug
2011-10-03 23:23 - 2.1.117 - normal udev
2011-10-03 23:36 - 2.1.118 - udev debug
2011-10-03 23:42 - 2.1.119 - udev debug
2011-10-04 14:48 - 2.1.120 - corrected bug with network_script
2011-10-04 18:28 - 2.1.121 - add atl1c
2011-10-04 18:41 - 2.1.122 -
2011-10-04 19:32 - 2.1.123 - rollback to 1.0.0.9 atl1e driver

ponix4-src/source/xorg-ponix4/app/transset-1.0.0/transSet.c
/* Copyright © 2003-2004 Matthew Hawn
 * Copyright © 2003-2004 Andreas Kohn
 * Copyright © 2003-2004 Roman Divacky
 * Copyright © 2003-2004 Keith Packard
 * Copyright © 2005-2007 Daniel Forchheimer
 * Copyright © 2011-2012 Arnaud Fontaine
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#define VERSIONSTR "6"
#define RELEASEDATESTR "2007-09-21"

#include <stdio.h>
#include <X11/Xlib.h>
#include <X11/Xatom.h>
#include "dsimple.h"
#include <stdlib.h>
#include <getopt.h>
#include <string.h>
#include <sys/types.h>

Window target_win;

/* needed by dsimple.c */

void
Usage (void)
{
    fprintf (stderr, "usage: transset-df [-options ...] [opacity]\n");
    fprintf (stderr, "options:\n");

    fprintf (stderr,
             "    -h, --help           display this message\n");
    fprintf (stderr,
             "    -t, --toggle         force toggle of opacity\n");
    fprintf (stderr,
             "    -c, --click          select by clicking on window  (default)\n");
    fprintf (stderr,
             "    -p, --point          select the window currently under the cursor\n");
    fprintf (stderr,
             "    -a, --actual         select the actual window\n");
    fprintf (stderr,
             "    -n, --name NAME      select by name, NAME is matched as regular expression\n");
    fprintf (stderr,
             "    --no-regex           don't use regular expression for matching name\n");
    fprintf (stderr,
             "    -i, --id             select by window id\n");
    fprintf (stderr,
             "        --inc            increase by the given opacity\n");
    fprintf (stderr,
             "        --dec            decrease by given opacity\n");
    fprintf (stderr,
             "    -m, --min OPACITY    minimum possible opacity  (default = 0)\n");
    fprintf (stderr,
             "    -x, --max OPACITY    maximum possible opacity  (default = 1)\n");
    fprintf (stderr,
             "    -v, --verbose        print some debug info\n");
    fprintf (stderr,
             "    -V, --version        print version number\n");

    exit (1);
}

#define OPAQUE	0xffffffff
#define OPACITY	"_NET_WM_WINDOW_OPACITY"

/* returns the highest parent of child that is not the root-window */
static Window
Get_Top_Window (Display *disp, Window child) {
    Window parent;
    Window root;
    Window *child_list;
    unsigned int num_children;

    if (!XQueryTree (disp, child, &root, &parent, &child_list,
                     &num_children))
        Fatal_Error ("Can't query window tree.");

    XFree ((void *) child_list);
    if (parent == root)
        return child;

    while (parent != root) {
        child = parent;
        if (!XQueryTree (disp, child, &root, &parent, &child_list,
                         &num_children))
            Fatal_Error ("Can't query window tree.");
        XFree ((void *) child_list);
    }
    return child;
}

static Window
Get_Actual_Window (Display *disp)
{
    int i;
    Window w;

    XGetInputFocus (disp, &w, &i);
    return Get_Top_Window (disp, w);
}

typedef enum {
    SELECT_METHOD_CLICK = 0,
    SELECT_METHOD_WINDOW_UNDER_CURSOR = 1,
    SELECT_METHOD_WINDOW_ID = 2,
    SELECT_METHOD_WINDOW_NAME = 3,
    SELECT_METHOD_FOCUSED_WINDOW = 4
} select_method_t;

int
main (int argc, char **argv)
{
    Bool gotd = False;
    double d = 0.75;
    unsigned int opacity;
    unsigned int current_opacity;
    select_method_t select_method = SELECT_METHOD_CLICK;
    Bool flag_toggle = False;
    Bool flag_increase = False;
    Bool flag_decrease = False;
    Bool flag_verbose = False;
    Bool flag_no_regex = False;
    int o;
    float min = 0, max = 1;
    char *idstr = NULL, *namestr = NULL;
    char *windowname = NULL;

    int options_index = 0;
    static struct option long_options[] = {
        {"toggle", 0, NULL, 't'},
        {"help", 0, NULL, 'h'},
        {"point", 0, NULL, 'p'},
        {"actual", 0, NULL, 'a'},
        {"click", 0, NULL, 'c'},
        {"id", 1, NULL, 'i'},
        {"name", 1, NULL, 'n'},
        {"inc", 0, NULL, '1'},
        {"dec", 0, NULL, '2'},
        {"min", 1, NULL, 'm'},
        {"max", 1, NULL, 'x'},
        {"no-regex", 0, NULL, '4'},
        {"version", 0, NULL, 'V'},
        {"verbose", 0, NULL, 'v'},
        {0, 0, 0, 0}
    };
    unsigned char *data;

    Atom actual;
    int format;
    unsigned long n, left;

    /* wonderful utility */
    Setup_Display_And_Screen (&argc, argv);

    /* parse arguments */
    while ((o = getopt_long (argc, argv, "thapci:n:vVm:x:123",
                             long_options, &options_index)) != -1) {
        switch (o) {
        case 't':
            flag_toggle = True;
            break;
        case 'h':
            Usage ();
            break;
        case 'c':
            select_method = SELECT_METHOD_CLICK;
            break;
        case 'p':
            select_method = SELECT_METHOD_WINDOW_UNDER_CURSOR;
            break;
        case 'i':
            idstr = optarg;
            select_method = SELECT_METHOD_WINDOW_ID;
            break;
        case 'n':
            namestr = optarg;
            select_method = SELECT_METHOD_WINDOW_NAME;
            break;
        case 'a':
            select_method = SELECT_METHOD_FOCUSED_WINDOW;
            break;
        case '1':
            flag_increase = True;
            break;
        case '2':
            flag_decrease = True;
            break;
        case 'v':
            flag_verbose = True;
            break;
        case '4':
            flag_no_regex = True;
            break;
        case 'm':
            min = atof (optarg);
            break;
        case 'x':
            max = atof (optarg);
            break;
        case 'V':
            fprintf (stderr, "version: %s\nreleased: %s\n",
                     VERSIONSTR, RELEASEDATESTR);
            exit (1);
            break;
        default:
            Usage ();
        }
    }

    if (optind < argc) {
        d = atof (argv[optind]);
        gotd = True;
    }

    /* select the window to make transparent */
    switch (select_method) {
    case SELECT_METHOD_WINDOW_UNDER_CURSOR:
        /* don't wait for click */
        if (flag_verbose)
            printf ("Selecting window by click\n");
        target_win = Get_Window_Under_Cursor (dpy);
        break;

    case SELECT_METHOD_WINDOW_ID:
        /* pretty much ripped from dsimple.c */
        if (flag_verbose)
            printf ("Selecting window by id\n");
        sscanf (idstr, "0x%lx", &target_win);
        if (!target_win)
            sscanf (idstr, "%ld", &target_win);
        if (!target_win) {
            fprintf (stderr, "Invalid window id format: %s.\n", idstr);
            XCloseDisplay (dpy);
            return 1;
        }
        if (flag_verbose)
            printf ("Selected 0x%x, trying to get top parent ... ",
                    (unsigned int) target_win);
        target_win = Get_Top_Window (dpy, target_win);
        if (flag_verbose)
            printf ("found 0x%x\n", (unsigned int) target_win);

        break;

    case SELECT_METHOD_WINDOW_NAME:
        /* select by name, pretty much ripped from dsimple.c */
        if (flag_verbose)
            printf ("Selecting window by name\n");

        if (flag_no_regex)
            target_win = Window_With_Name (dpy, RootWindow (dpy, screen),
                                           namestr);
        else
            target_win = Window_With_Name_Regex (dpy, RootWindow (dpy, screen),
                                                 namestr);

        if (!target_win) {
            fprintf (stderr, "No window matching %s exists!\n", namestr);
            XCloseDisplay (dpy);
            return 1;
        }
        /* store the matched window name*/
        XFetchName (dpy, target_win, &windowname);

        if (flag_verbose)
            printf ("Selected 0x%x, trying to get top parent ... ",
                    (unsigned int) target_win);

        target_win = Get_Top_Window (dpy, target_win);
        if (flag_verbose)
            printf ("found 0x%x\n", (unsigned int) target_win);

        break;

    case SELECT_METHOD_FOCUSED_WINDOW:
        target_win = Get_Actual_Window (dpy);
        break;

    default:
        /* grab mouse and return window that is next clicked */
        target_win = Select_Window (dpy);
        break;
    }

    /* get property */
    XGetWindowProperty (dpy, target_win, XInternAtom (dpy, OPACITY, False),
                        0L, 1L, False, XA_CARDINAL, &actual, &format, &n, &left,
                        (unsigned char **) &data);

    if (data != None) {
        memcpy (&current_opacity, data, sizeof (unsigned int));
        XFree ((void *) data);
        if (flag_verbose)
            printf ("Found transparency: %g\n",
                    (double) opacity / OPAQUE);
    } else
        current_opacity = OPAQUE;

    if (flag_increase)
        d = (double) current_opacity / OPAQUE + d;
    else if (flag_decrease)
        d = (double) current_opacity / OPAQUE - d;

    /* check min and max */
    if (d < min)
        d = min;
    if (d > max)
        d = max;

    opacity = (unsigned int) (d * OPAQUE);

    /* for user-compability with transset */
    if (!gotd)
        flag_toggle = True;

    /* toggle */
    if (flag_toggle)
        if (current_opacity != OPAQUE)
            opacity = OPAQUE;

    if (opacity == OPAQUE)
  	XDeleteProperty (dpy, target_win, XInternAtom (dpy, OPACITY, False));
    /* set it */
    else
        XChangeProperty (dpy, target_win, XInternAtom (dpy, OPACITY, False),
                         XA_CARDINAL, 32, PropModeReplace,
                         (unsigned char *) &opacity, 1L);
    XSync (dpy, False);

    printf ("Set Property to %g", (double) opacity / OPAQUE);
    if (windowname)
        printf (" on \n%s\n", windowname);
    else
        printf ("\n");

    if (flag_verbose)
        printf ("Propery set on: 0x%x\n", (unsigned int) target_win);

    XCloseDisplay (dpy);
    return 0;
}
